### ISSUE TITLE
cmake fails when a component of the path to unpacked SuiteSparse tarball matches "[0-9]+.[0-9]+.[0-9]+" regex

### ISSUE THREAD

**Issue by balducci** [DrTimothyAldenDavis/SuiteSparse#154]
hello

**Describe the bug**
cmake configuration of SuiteSparse-6.0.0-beta1 fails when SuiteSparse
is unpacked under a path already containing a component matching the
``[0-9]+.[0-9]+.[0-9]+`` regex.

For instance, when the whole path to the unpacked tarball is:
```
/home/balducci/tmp/install-us-d/foo-1.2.3/SuiteSparse-6.0.0-beta1
```

the build chokes almost immediately for me with the cmake error
message:
```
CMake Error at /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find SuiteSparse_config: Found unsuitable version "1.2.3", but
  required is at least "6.0.0" (found
  /home/balducci/tmp/install-us-d/foo-1.2.3/SuiteSparse-6.0.0-beta1/SuiteSparse_config/build/libsuitesparseconfig.so.6.0.0)
```

AFAICS, this is due to the ``string( REGEX MATCH "[0-9]+.[0-9]+.[0-9]+" ...)``
calls in the ``SuiteSparse_config/cmake_modules/*.cmake``modules, which
return upon the first match (while they shouldn't)

Anchoring the regex match to the end of the string, like in:
```
string( REGEX MATCH "[0-9]+.[0-9]+.[0-9]+[^0-9]*$" ...)
```
fixes things for me

**To Reproduce**
```
mkdir /tmp/foo-1.2.3
cd /tmp/foo-1.2.3
tar axf /location/of/the/SuiteSparse-6.0.0-beta1.tar.gz
cd SuiteSparse-6.0.0-beta1
make library
```

**Desktop (please complete the following information):**
 - OS: ``Linux 6.0.3 #1 SMP Sat Oct 22 07:54:54 CEST 2022 x86_64 GNU/Linux``
 - compiler:
 ```
gcc (GCC) 12.2.0
Copyright (C) 2022 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```
 - BLAS and LAPACK library:    ``openblas-0.3.21``

thank you very much for your valuable work
ciao
-gabriele

**Comment by DrTimothyAldenDavis** [comment]
Thanks for the feedback!  I just tried it but the solution breaks on my Mac, because the libraries are named libamd.3.0.0.dylib (for example), which differs from the naming scheme in Linux (libamd.so.3.0.0).  So the "$" to anchor it on the end of the string doesn't work.

It strikes me that there should be standard cmake way to query an installed library for its version number.  I have to go through terrible gyrations to get the MPFR and GMP version numbers.  For example:

https://github.com/DrTimothyAldenDavis/SuiteSparse/blob/master/SuiteSparse_config/cmake_modules/FindGMP.cmake

And how would this work on Windows?  Yikes ... I have no way of testing that.

I'll dig into the REGEX command to see if it can match the *last* occurence instead of the first.  Maybe that will be a portable solution, instead of adding "$".

**Comment by balducci** [comment]
> Thanks for the feedback! I just tried it but the solution breaks on my Mac, because the libraries are named libamd.3.0.0.dylib (for example), which differs from the naming scheme in Linux (libamd.so.3.0.0). So the "$" to anchor it on the end of the string doesn't work.

wait: the anchoring  is obtained not simply adding a ``$``, but   ``[^0-9]*$``, which should happily catch ``3.0.0`` in, eg:

``.../foo-1.2.3/SuiteSparse-6.0.0-beta1/.../libamd.3.0.0.dylib``

no?


> I'll dig into the REGEX command to see if it can match the _last_ occurence instead of the first. Maybe that will be a portable solution, instead of adding "$".

There is a ``REGEX MATCHALL`` variant:

``Match the <regular_expression> as many times as possible and store the matches in the <output_variable> as a list``

Maybe you could take advantage of that...

ciao
-g

**Comment by DrTimothyAldenDavis** [comment]
That works ... sort of.  I get the version numbers as "6.0.0.dylib" on the Mac.  That works but it's ugly.

I think I'll use another approach.

file ( UMFPACK_FILENAME ${UMFPACK_LIBRARY} NAME )

returns the filename without the whole path.  There should be only one X.Y.X in that filename.  Then I can use the simple REGEX MATCH I had before to find that one instance in the filename.

Only downside to all of these methods ... I have no idea if this works on Windows because cmake doesn't have a standard method for querying version numbers X.Y.Z from a compiled library.    Argh ...

**Comment by DrTimothyAldenDavis** [comment]
This should be fixed in the v6.0.0-beta4 version:
https://github.com/DrTimothyAldenDavis/SuiteSparse/releases/tag/v6.0.0-beta4

can you give it a try and let me know if it resolves this issue for you, too?  I've tested it by renaming my SuiteSparse folder temporarily to SuiteSparse-7.0.0 (a fake version), and the new REGEX and file (... NAME) process works fine with that revision, correctly ignoring the "7.0.0" in the path and finding the correct X.Y.Z version number.

**Comment by balducci** [comment]
> This should be fixed in the v6.0.0-beta4 version: https://github.com/DrTimothyAldenDavis/SuiteSparse/releases/tag/v6.0.0-beta4
> 
> can you give it a try and let me know if it resolves this issue for you, too? 

confirmed: v6.0.0-beta4 builds fine and the reported problem is fixed

so closing this issue

thank you very much again!
